### ISSUE TITLE
Errors in ATACorrect 

### ISSUE THREAD

**Issue by Danvor** [loosolab/TOBIAS#14]
I keep getting the following for all of my regions in the bed file when running: 
# Command line call: TOBIAS ATACorrect --bam merged_ATAC.bam --genome /home/danvor/Data/genome.fa.gz --peaks /home/danvor/Data/putative_CRMs.bed --outdir atac_corrections/ --prefix atac --cores 16

2020-04-22 13:15:30 (21235) [ERROR]     Error writing key: corrected:both, region: ('Scaffold31218', 382, 631) to bigwig
2020-04-22 13:15:30 (21235) [ERROR]     Error writing key: corrected:both, region: ('Scaffold31230', 62, 524) to bigwig
2020-04-22 13:15:30 (21235) [ERROR]     Error writing key: corrected:both, region: ('Scaffold31376', 356, 832) to bigwig
2020-04-22 13:15:30 (21235) [ERROR]     Error writing key: corrected:both, region: ('Scaffold31601', 62, 798) to bigwig
2020-04-22 13:15:30 (21235) [ERROR]     Error writing key: corrected:both, region: ('Scaffold31683', 260, 615) to bigwig
2020-04-22 13:15:30 (21235) [ERROR]     Error writing key: corrected:both, region: ('Scaffold31710', 62, 384) to bigwig

The resulting atac_corrected.bw and  atac_uncorrected.bw are the same and the md5sum values are the same. I don't get any other errors in the log except the above.

Are there any requirements for the bed file (like number of columns) or for bed?
I have used the same genome for mapping and peak calling, they definitely have compatible Scaffold/Chromosome IDs.

But I also get the same error when using the test bam, genome and peak files, though.

Thanks.
Danvor

**Comment by Danvor** [comment]
I get the same error when using the test bam, genome and peak files, though.

**Comment by msbentsen** [comment]
Thanks for sending this! Although it runs through, I guess the .bw-files are empty since no keys could be written?

I think it is something similar to #10 related to the pybigwig library used for reading/writing files. Can you run `pip freeze | grep pyBigWig` for me to get the version number?

**Comment by Danvor** [comment]
Hi, 
Thanks for the reply. I think we don't have pip on the server, so here is what I managed to find.

/opt/pyenv/tobias-env/lib/python3.8/site-packages/pyBigWig-0.3.17-py3.8.egg-info
/opt/pyenv/tobias-env/lib/python3.8/site-packages/pyBigWig-0.3.17-py3.8.egg-info/PKG-INFO
/opt/pyenv/tobias-env/lib/python3.8/site-packages/pyBigWig-0.3.17-py3.8.egg-info/SOURCES.txt
/opt/pyenv/tobias-env/lib/python3.8/site-packages/pyBigWig-0.3.17-py3.8.egg-info/dependency_links.txt
/opt/pyenv/tobias-env/lib/python3.8/site-packages/pyBigWig-0.3.17-py3.8.egg-info/installed-files.txt
/opt/pyenv/tobias-env/lib/python3.8/site-packages/pyBigWig-0.3.17-py3.8.egg-info/requires.txt
/opt/pyenv/tobias-env/lib/python3.8/site-packages/pyBigWig-0.3.17-py3.8.egg-info/top_level.txt

I hope this is helpful.
Best,
Danvor

**Comment by msbentsen** [comment]
That helped a lot, thanks! I can reproduce the error for that (and some other) versions of pyBigWig, which seems to be due to changes in the API for adding values to bigwigs. I have pushed a few changes to the development branch of TOBIAS which should fix this.

Can you install using the commands below and confirm whether the problem is solved for you? 
```
git clone -b dev https://github.com/loosolab/TOBIAS.git
cd TOBIAS
python setup.py install
```
`TOBIAS --version` should yield `0.11.1-dev`.

**Comment by Danvor** [comment]
Hi, 
Thanks a lot. This fixed the error, however, it seems that now there is a related error in the following command, that uses the output of ATACorrect on the test data you provided: 

## Command line call: TOBIAS ScoreBigwig --signal tob_test/test_corrected.bw --regions merged_peaks.bed --output test_footprints.bw --cores 16

2020-04-24 13:07:57 (25050) [INFO]      Processing input files
2020-04-24 13:07:57 (25050) [INFO]      - Opening input cutsite bigwig
2020-04-24 13:07:57 (25050) [INFO]      - Getting output regions ready
2020-04-24 13:07:57 (25050) [INFO]      Calculating footprints in regions...
2020-04-24 13:07:57 (25050) [INFO]      Progress 0%
2020-04-24 13:08:00 (25050) [INFO]      Progress 1.0%
2020-04-24 13:08:00 (25050) [INFO]      Progress 2.0%
2020-04-24 13:08:00 (25050) [INFO]      Progress 3.0%
2020-04-24 13:08:00 (25050) [INFO]      Progress 4.0%
2020-04-24 13:08:00 (25050) [INFO]      Progress 5.0%
2020-04-24 13:08:01 (25050) [INFO]      Progress 8.0%
2020-04-24 13:08:01 (25050) [INFO]      Progress 13.0%
2020-04-24 13:08:01 (25050) [INFO]      Progress 14.0%
2020-04-24 13:08:01 (25050) [INFO]      Progress 16.0%
Error reading region: ('chr4', 30989470, 30989965, '.') from pybigwig object
Error reading region: ('chr4', 31045662, 31046133, '.') from pybigwig object
Error reading region: ('chr4', 31234339, 31234976, '.') from pybigwig object
Error reading region: ('chr4', 31259121, 31259748, '.') from pybigwig object
Error reading region: ('chr4', 31275253, 31275713, '.') from pybigwig object
Error reading region: ('chr4', 31290683, 31291757, '.') from pybigwig object
Error reading region: ('chr4', 31370778, 31371324, '.') from pybigwig object
Error reading region: ('chr4', 31390788, 31391293, '.') from pybigwig object
Error reading region: ('chr4', 31530813, 31531326, '.') from pybigwig object
Error reading region: ('chr4', 31538889, 31539543, '.') from pybigwig object
Error reading region: ('chr4', 31674260, 31675053, '.') from pybigwig object
Error reading region: ('chr4', 31737268, 31738112, '.') from pybigwig object
and so on.

Thanks.
Best,
Danvor

**Comment by Danvor** [comment]
It may be worth mentioning though that for some reason the md5sum of the output of ATACorrect ran on the test data is not the same as your output file.
Provided test output:
6991db024c0095376552243a10590f5e  Bcell_corrected.bw
9608d377166e703f48ecaafdf4729a50  Bcell_uncorrected.bw
Resulting test output:
3867fd18b32207e31779f359948e138a  tob_test/test_corrected.bw
9608d377166e703f48ecaafdf4729a50  tob_test/test_uncorrected.bw

Thanks.
Danvor

**Comment by msbentsen** [comment]
Hi Danvor,

I had a look at the issue and the responsibility is entirely on me! I had falsely assumed that `pyBigWig` would always be compiled with numpy, leading to errors when that was not the case. Could you pull from the dev-branch once more to confirm the fix? 

Regarding the md5sums, I don't have a good answer for why they are different. Could be that different versions of pyBigWig (or other packages) affect the compression of the .bw's, or that there are other system-specific influences. Running `TOBIAS ATACorrect` twice with the same data yields the same md5sums (e.g. no randomness), so I would not worry about differences in the content of the .bigwigs. Although I do understand the annoyance when trying to compare the output.

Best Mette

**Comment by Danvor** [comment]
Hi, yes, thank you. This works as expected now.

**Comment by msbentsen** [comment]
Great, thank you for testing it! I have just released `0.11.1` with these changes, so that should solve this and other issues with reading/writing bigwigs.
